distributed/semaphore.py

```diff
@@ -0,0 +1,264 @@
+                    self._release_value, name=name, client=client, identifier=_id
+                )
+
+    @gen.coroutine
+    def _validate_leases(self):
+        if not self._validation_running:
+            self._validation_running = True
+            known_clients = set(self.leases_per_client.keys())
+            scheduler_clients = set(self.scheduler.clients.keys())
+            for client in known_clients - scheduler_clients:
+                client_has_leases = sum(
+                    valmap(len, self.leases_per_client[client]).values()
+                )
+                if client_has_leases:
+                    self._release_client(client)
+            else:
+                self._validation_running = False
+
+
+class Semaphore(object):
+    def __init__(self, max_leases=1, name=None, client=None):
+        self.client = client or _get_global_client() or get_worker().client
+        self.id = uuid.uuid4().hex
+        self.name = name or "semaphore-" + uuid.uuid4().hex
+        self.max_leases = max_leases
+
+        if self.client.asynchronous:
+            self._started = self.client.scheduler.semaphore_create(
+                name=self.name, max_leases=max_leases
+            )
+        else:
+            self.client.sync(
+                self.client.scheduler.semaphore_create,
+                name=self.name,
+                max_leases=max_leases,
+            )
+            self._started = gen.moment
+
+    def __await__(self):
+        @gen.coroutine
+        def _():
+            yield self._started
+            raise gen.Return(self)
+
+        return _().__await__()
+
+    def acquire(self, timeout=None):
+        """
+        Acquire a semaphore.
+
+        If the internal counter is greater than zero, decrement it by one and return True immediately.
+        If it is zero, wait until a release() is called and return True.
+        """
+        return self.client.sync(
+            self.client.scheduler.semaphore_acquire,
+            name=self.name,
+            timeout=timeout,
+            client=self.client.id,
+            identifier=self.id,
+        )
+
+    def release(self):
+        """
+        Release a semaphore.
+
+        Increment the internal counter by one.
+        """
+
+        """ Release the lock if already acquired """
+        return self.client.sync(
+            self.client.scheduler.semaphore_release,
+            name=self.name,
+            client=self.client.id,
+            identifier=self.id,
+        )
+
+    def __enter__(self):
+        self.acquire()
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.release()
+
+    @gen.coroutine
+    def __aenter__(self):
+        yield self.acquire()
+        raise gen.Return(self)
+
+    @gen.coroutine
+    def __aexit__(self, *args, **kwargs):
+        yield self.release()
+
+    def __getstate__(self):
+        return (self.name, self.client.scheduler.address, self.max_leases)
```

Comment:
Should `id` also be serialized here, or do we want `Semaphore`s which have gone through the serialization / deserialization process to be considered different?

Reply:
TL;DR: I think we should not serialize the ID. I should probably add a test for this behavior.

Consider the following:

```python
from dask import delayed
from distributed import Semaphore

def access_database(sem, *args):
    with sem:
        ...
    return data

sem = Semaphore("my_database", max_leases=2)
tasks = []
for ix in range(10):
    tasks.append(
        delated(access_database)(sem, ix)
    )

dask.compute(tasks)
```

This is how I would probably write a piece of code to download data from a DB. I could rely on `dask.distributed` to serialize the class for me. Otherwise I would need to pass all initialization parameters to the function and initialize the objects myself to keep everything consistent.
When I use the serialization approach, I still require the semaphore instance for every job to be unique by ID. Otherwise the tracking is off and the mechanism for tracking leases by ID and client would not work.